### ISSUE TITLE
GTP fix 2TB-limit readynas duo v1

### ISSUE THREAD

**Issue by mfe-** [mfe-/ReadyNASDuoSparc#2]
@JustLittleTinyMe is looking for a 2TB-limit fix.
As I don't know much about this topic, maybe someone more competent can jump in?

What work needs to be done to support HDD with more then 2TB? Is it only be solve able by upgrading the bootloader?

**Comment by JustLittleTinyMe** [comment]
The way I understand it, the bootloader loads the Kernel, which in turn can load modules containing support for, lets say, for instance - GPT. 

Meaning (_IF_ I'm right about it), one gets the kernel to load an additional module with support for GPT, and the issue should pretty much be solved. Perhaps there are additional scripts that needs to be edited considering partitioning etc, but essentially, the only thing limiting an OS from reaching beyond 2TB, is the MBR-partitioning-scheme. 

That is - _IF_ the SATA-chipset allows for larger drives. Which I have found no information about what so ever. The only reference from NetGear is considering the SPARC-processor, not the SATA-interface. Meaning, the way I read it, it's a software-issue only. 

Or, what have I missed?

This is the closest to an answer I've found so far: 
https://community.netgear.com/t5/New-ReadyNAS-Users-General/ReadyNAS-DUO-v-1-amp-4TB-is-that-possible/td-p/956138

**Comment by billsargent** [comment]
If it were that easy, it would most likely have been done before now. I suspect the chipset itself doesn't support that. At these devices EOL, there had already been larger than 2TB drives available. It would be better to simply migrate away from these decaying old NAS units and to something faster and more reliable. The sparc variant totally pissed me off with its odd instruction set that won't allow normal GCC compiled stuff unless gcc was patched with some odd patches. It's insane they ever released the nas with such a stupid modified CPU. I'm on Asustor these days.

**Comment by JustLittleTinyMe** [comment]
And exactly therefore I'm sure noone has ever tried doing it. Not even the guys at NetGear put a single hour on deprecated tech. Especially not if they have to do some extra patching/configuring of the compiler before they even can begin.

**Comment by billsargent** [comment]
Yeah. I mean I really really wanted to upgrade Samba on it and a few other paackages, but since the GCC version is so old, nothing would work. I tried building a new patched GCC and it just failed. Too many differences between things.  My new idea is to build a new nas out of the enclosure using a small development board that has a SATA controller on it. It might not be faster than the sparc processor in mine, but it'll work. I just need to get a normal power supply for it since the backplane will serve no use anymore. :)

**Comment by mfe-** [comment]
> Yeah. I mean I really really wanted to upgrade Samba on it and a few other paackages, but since the GCC > version is so old, nothing would work. I tried building a new patched GCC and it just failed.

Doesn't work the provided gcc 4,6 or gcc 7.1 in this repository for you? Which error did you get?

**Comment by billsargent** [comment]
No, those don't work for most things. you'll get illegal instruction or segfaults. and you most definately can't compile a kernel or module with anything new. As I said, the sparc processor on these are non-standard. I'd been working with a few people to try to patch a newer GCC and libc to compile a newer kernel, but it won't boot.  Now let me rephrase some of this... the newer gcc will work on some things, but for example when compiling samba, samba requires a lot of newer utilities, and libraries that can't be compiled properly with newer GCC. You're stuck with gcc 3.x patched for the infrant sparc processor. It was a stupid design by infrant.

**Comment by billsargent** [comment]
Go down to the gotchas section of this http://wiki.dietpc.org/wiki/DIET-PC_on_SPARC_ReadyNAS

**Comment by Netfreak25** [comment]
I was just messing around with the old nas and it looks like there is a fix. I am not sure as I can't test it as I have 2TB disks installed, but worth a try if you have bigger disks. It's an addon

https://kb.netgear.com/24546/Add-ons-for-RAIDiator-4-1-3-Sparc

"ForceExpansion - Data volume size greater than 2TB is now supported. Existing X-RAID systems utilizing 750GB will not automatically expand . you will need to use the ForceExpansion add-on to reclaim the extra 40GB of space beyond the previous pre-RAIDiator 4.0 2TB limitation. Note that if you will be using 1TB drives, make sure to check the Hard Disk Compatibility List."

**Comment by mfe-** [comment]
@Netfreak25 Thanks for letting us know. Does anyone else have experience with this addon? From the description I'm not sure if you receive the total amount of space of the HDD or only 40GB.  Currently I don't have any free disks above 2TB so, I guess I have to wait too.

**Comment by emeyedeejay** [comment]
Also messing around with my old NAS Netgear ReadyNas Duo ...

Installed 2 x WD40EZRX 4TB drives but only 1667MB is seen by the NAS on each disk.

I installed the above mentioned ForceExpansion add-on but it made no difference

![image](https://user-images.githubusercontent.com/8543879/145538305-0e6f14bd-57c5-40ee-91c4-3ff61caa1ccb.png)

![image](https://user-images.githubusercontent.com/8543879/145538324-ce8625e9-ea0b-4b4c-a774-8f657982354e.png)

I'm not sure why they would report as different available sizes.

**Comment by billsargent** [comment]
I really believe this is a limitation of the arch, not a software fixable issue. Its about addressing the space hardware wise.

**Comment by robmck** [comment]
Coincidentally, I also just tried this yesterday by stuffing four 4Tb WD Red+ drives in my ReadyNAS NV+, doing Factory Reset, then adding Force Grow add-on. It only saw 1.7Tb of each drive, thus creating a ~5Tb volume. 

Installing root ssh, apt, and parted, I looked around the partitions. Here's what it did on each of the drives:

```
Disk geometry for /dev/hdg: 0.000-1718295.835 megabytes
Disk label type: msdos
Minor    Start       End     Type      Filesystem  Flags
1          0.016   2000.015  primary   ext3        raid
2       2000.016   2512.015  primary               raid
```

I tried creating a gpt partition table, but that didn't enable any more address space. I notice at the top that parted is only seeing disk geometry for the first 1.7Tb. Something below parted is the problem.

I couldn't find a way to see if the kernel had gpt support compiled in. I tried a dist-upgrade a few times, hoping a newer distro would have it turned on.  Unfortunately, that always left it unresponsive (no access to admin site, or ssh, or much else other than the boot menu so I could do an OS Install). 

I could go try to build my own kernel, but as I don't know if the boot firmware or hardware supports this. Building my own kernel could be a big waste of time. At this point, it's easier to buy a new ReadyNAS, Synology, or just stuff the drives into one of my old PCs and install truenas on it.

**Comment by emeyedeejay** [comment]
Yeah - I'm leaving this alone now. This NAS went eol in 2011 I think I read somewhere...

I have it running with 2x4TBs each of which the NAS sees the same 1.7TB you mention. I enabled NFS and am able to mount NFS shares on my proxmox server so I could use that for backup but seems a waste of the other 2.7TB of disk space.

And yeah - old PC, TrueNAS and done is probably the way I'll go too 😁

**Comment by mfe-** [comment]
I just added gptfdisk-1.0.8_1.0.8-1_sparc.deb with d631df264897579926faadcc5dd98afd12fb9b12 . It does not address the 2TB limit but maybe if someone wants to play around with it. Maybe we can create 4x1 TB partitions such that a 4TB Hdd can be used?
